File: app/src/test/java/com/example/mike/mp3player/client/TimerUtilsTest.java
package com.example.mike.mp3player.client;

import android.os.SystemClock;
import android.support.v4.media.session.PlaybackStateCompat;
import android.util.Log;

import com.example.mike.mp3player.client.utils.TimerUtils;

import org.junit.Test;
import org.junit.jupiter.api.BeforeAll;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowSystemClock;

import static org.junit.Assert.assertEquals;


/**
 * In order to use Powermockito, used old Test tag so that the class runs with JUnit4.
 */
@RunWith(RobolectricTestRunner.class)
@Config(manifest=Config.NONE, sdk = 26, shadows = {ShadowSystemClock.class})
public class TimerUtilsTest {
    /**
     * Class setup method
     */
    @BeforeAll
    public static void setupTests() {
        final long currentTime = SystemClock.elapsedRealtime();
        String wrongCurrentTimeErrorMessage = "SystemClock.elapsedRealTime() (i.e. ShadowSystemClock)" +
                " returned unexpected result. Tests are based on the assumption that this method will" +
                " return" + SHADOW_SYSTEM_CLOCK_ELAPSED_TIME +
                ". Consult the Robolectric documentation for more information";
        assertEquals(wrongCurrentTimeErrorMessage, SHADOW_SYSTEM_CLOCK_ELAPSED_TIME, currentTime);
    }
    /**
     * assumed elapsed time will return 100L.
     */
    private static final long SHADOW_SYSTEM_CLOCK_ELAPSED_TIME = 100L;

    @Test
    public void testConvertToSecondsOneSecond() {
        long milliseconds = 1000;
        int result = TimerUtils.convertToSeconds(milliseconds);
        int expect = 1;
        assertEquals(expect, result);
    }

    @Test
    public void testConvertToSecondsMillisecondsToRound() {
        long milliseconds = 17972;
        int result = TimerUtils.convertToSeconds(milliseconds);
        int expect = 17;
        assertEquals(expect, result);
    }

    @Test
    public void testFormatTimeOneSecond() {
        long milliseconds = 1000;
        String result = TimerUtils.formatTime(milliseconds);
        String expect = "00:01";
        assertEquals(expect, result);

    }
    @Test
    public void testFormatTimeRoundableTime() {
        long milliseconds = 78756;
        String result = TimerUtils.formatTime(milliseconds);
        String expect = "01:18";
        assertEquals(expect, result);
    }
    @Test
    public void calculateCurrentPlaybackPositionWhenStatePlayingTest() {
        final long timeDiff = 400L;
        final long originalTime = SystemClock.elapsedRealtime() - timeDiff;
        final long originalPosition = 40000L;
        PlaybackStateCompat playbackStateCompat =
                new PlaybackStateCompat.Builder().setState(PlaybackStateCompat.STATE_PLAYING,
                        originalPosition, 0f, originalTime).build();
        final long newPosition = TimerUtils.calculateCurrentPlaybackPosition(playbackStateCompat);
        final long expectedNewPosition = originalPosition + timeDiff;
        assertEquals(expectedNewPosition, newPosition);
    }

    @Test
    public void calculateCurrentPlaybackPositionWhenNotStatePlayingTest() {
        PlaybackStateCompat playbackStateCompat =
                new PlaybackStateCompat.Builder().setState(PlaybackStateCompat.STATE_PAUSED, 40000L, 0f, 0).build();
        long newPosition = TimerUtils.calculateCurrentPlaybackPosition(playbackStateCompat);
        assertEquals(playbackStateCompat.getPosition(), newPosition);
    }


}

File: app/src/test/java/com/github/goldy1992/mp3player/client/TimerUtilsTest.kt
package com.github.goldy1992.mp3player.client

import android.os.SystemClock
import android.support.v4.media.session.PlaybackStateCompat
import com.github.goldy1992.mp3player.client.utils.TimerUtils.calculateCurrentPlaybackPosition
import com.github.goldy1992.mp3player.client.utils.TimerUtils.convertToSeconds
import com.github.goldy1992.mp3player.client.utils.TimerUtils.formatTime
import org.junit.Assert
import org.junit.Test
import org.junit.jupiter.api.BeforeAll
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowSystemClock

/**
 * Test class for the TimerUtils class
 */
@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE, sdk = [26], shadows = [ShadowSystemClock::class])
class TimerUtilsTest {
    @Test
    fun testConvertToSecondsOneSecond() {
        val milliseconds: Long = 1000
        val result = convertToSeconds(milliseconds)
        val expect = 1
        Assert.assertEquals(expect.toLong(), result.toLong())
    }

    @Test
    fun testConvertToSecondsMillisecondsToRound() {
        val milliseconds: Long = 17972
        val result = convertToSeconds(milliseconds)
        val expect = 17
        Assert.assertEquals(expect.toLong(), result.toLong())
    }

    @Test
    fun testFormatTimeOneSecond() {
        val milliseconds: Long = 1000
        val result = formatTime(milliseconds)
        val expect = "00:01"
        Assert.assertEquals(expect, result)
    }

    @Test
    fun testFormatTimeRoundableTime() {
        val milliseconds: Long = 78756
        val result = formatTime(milliseconds)
        val expect = "01:18"
        Assert.assertEquals(expect, result)
    }

    @Test
    fun calculateCurrentPlaybackPositionWhenStatePlayingTest() {
        val timeDiff = 400L
        val originalTime = SystemClock.elapsedRealtime() - timeDiff
        val originalPosition = 40000L
        val playbackStateCompat = PlaybackStateCompat.Builder().setState(PlaybackStateCompat.STATE_PLAYING,
                originalPosition, 0f, originalTime).build()
        val newPosition = calculateCurrentPlaybackPosition(playbackStateCompat)
        val expectedNewPosition = originalPosition + timeDiff
        Assert.assertEquals(expectedNewPosition, newPosition)
    }

    @Test
    fun calculateCurrentPlaybackPositionWhenNotStatePlayingTest() {
        val playbackStateCompat = PlaybackStateCompat.Builder().setState(PlaybackStateCompat.STATE_PAUSED, 40000L, 0f, 0).build()
        val newPosition = calculateCurrentPlaybackPosition(playbackStateCompat)
        Assert.assertEquals(playbackStateCompat.position, newPosition)
    }

    companion object {
        /**
         * Class setup method
         */
        @BeforeAll
        fun setupTests() {
            val currentTime = SystemClock.elapsedRealtime()
            val wrongCurrentTimeErrorMessage = "SystemClock.elapsedRealTime() (i.e. ShadowSystemClock)" +
                    " returned unexpected result. Tests are based on the assumption that this method will" +
                    " return" + SHADOW_SYSTEM_CLOCK_ELAPSED_TIME +
                    ". Consult the Robolectric documentation for more information"
            Assert.assertEquals(wrongCurrentTimeErrorMessage, SHADOW_SYSTEM_CLOCK_ELAPSED_TIME, currentTime)
        }

        /**
         * assumed elapsed time will return 100L.
         */
        private const val SHADOW_SYSTEM_CLOCK_ELAPSED_TIME = 100L
    }
}